src/MiddlewareFactory.php

```diff
@@ -0,0 +1,147 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
+ * @copyright Copyright (c) 2018 Zend Technologies USA Inc. (https://www.zend.com)
+ * @license   https://github.com/zendframework/zend-expressive/blob/master/LICENSE.md New BSD License
+ */
+
+declare(strict_types=1);
+
+namespace Zend\Expressive;
+
+use Psr\Container\ContainerInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Zend\Stratigility\MiddlewarePipe;
+use Zend\Stratigility\Middleware\CallableMiddlewareDecorator;
+use Zend\Stratigility\Middleware\PathMiddlewareDecorator;
+
+/**
+ * Marshal middleware for use in the application.
+ *
+ * This class provides a number of methods for preparing and returning
+ * middleware for use within an application.
+ *
+ * If any middleware provided is already a MiddlewareInterface, it can be used
+ * verbatim or decorated as-is. Other middleware types acceptable are:
+ *
+ * - string service names resolving to middleware
+ * - arrays of service names and/or MiddlewareInterface instances
+ * - PHP callables that follow the PSR-15 signature
+ *
+ * Additionally, the class provides several decorator/utility methods:
+ *
+ * - callable() will decorate the callable middleware passed to it using
+ *   CallableMiddlewareDecorator.
+ * - path() will decorate the path and middleware passed to it using
+ *   PathMiddlewareDecorator, after first passing the middleware to the prepare()
+ *   method.
+ * - pipeline() will create a MiddlewarePipe instance from the array of
+ *   middleware passed to it, after passing each first to prepare().
+ */
+class MiddlewareFactory
+{
+    /**
+     * @var null|MiddlewareContainer
+     */
+    private $container;
+
+    public function __construct(ContainerInterface $container = null)
+    {
+        $this->container = $container ? new MiddlewareContainer($container) : null;
```

Comment:
For better testability we should pass `MiddlewareContainer` as param instead of creating one in the constructor.

Reply:
It makes usability harder, however, as developers who are creating an instance on the fly now need to do the following:

```php
$factory = new MiddlewareFactory(new MiddlewareContainer($app->getContainer()));
```

Additionally, the above example I present would fail with an `E_FATAL` if `$app->getContainer()` returns `null` (which is possible currently), which means it technically needs to become:

```php
$factory = new MiddlewareFactory(
    $app->getContainer()
    ? new MiddlewareContainer($app->getContainer())
    : null
);
```

which is not user friendly at all. (FTR, if the middleware in a specification or a pipeline resolves to an constructor-less class, a callable, or a `MiddlewareInterface` instance, the factory can still do its work; it's only if a service name is used that the container becomes necessary. At that point, it raises an exception when a container is missing.)

Reply:
I've since changed this to accept only a `MiddlewareContainer` instance.